2-Draft-Accepted/RFCNNNN-Support-Experimental-Features.md

```diff
@@ -0,0 +1,626 @@
+---
+RFC: RFC
+Author: Dongbo Wang, Dan Travison
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Engine and Configuration
+Comments Due: March 22nd, 2018
+Plan to implement: Yes
+---
+
+# Support Experimental Features via Configuration
+
+This RFC is about adding the Experimental Feature support in PowerShell through the PowerShell configuration.
+
+The Experimental Feature support in PowerShell is to provide a mechanism for experimental features to coexist
+with existing stable features in PowerShell or PowerShell modules.
+An experimental feature would not be enabled
+unless a user opt in by properly configuring the `powershell.config.json` file.
+Experimental features should not affect any stable features
+and not present any side effects
+when they are all disabled.
+
+## Motivation
+
+This is a highly demanded feature which we believe will vastly improve the productivity of
+PowerShell developers on experimenting or iterating new features in PowerShell and PowerShell modules.
+
+The Experimental Feature support would make it easy for users to try out features
+that are in early or intermediate stages of development,
+and thus, allow the developers to get early feedback
+and make adjustments to the design or implementation based on the feedback.
+
+For users that do not care about the new features,
+they can continue to enjoy the bug fixes and other improvements from the new releases
+without worrying about regressions, issues, or side effects caused by the new features.
+
+For example, PowerShell team plans to rewrite the `FileSystemProvider` in future
+to reduce the code complexity and improve the performance.
+During the development,
+the `FileSystemProviderV2` can coexist with the existing one
+and be exposed to the user as an experimental feature,
+so that PowerShell team can get early feedback and issue reports.
+At the same time, users can continue to use the existing `FileSystemProvider`
+without being affected by the known or unknown issues introduced by the `FileSystemProviderV2`.
+
+## Specification
+
+### Experimental Feature Names
+
+Experimental features for PowerShell components and built-in modules should be named as `PSXXX`,
```

Comment:
I am thinking about easy searching in code sources - what about `PSEX` or `PSEXP`?

Reply:
I'm open to `PSEXP`.

Reply:
Had a second thought about the naming. I want to avoid using a string of feature name for existence test in engine code, like `if (context.EnabledExpFeatures.Contains("PSEXPBlahFeature"))`. Directly using a string for checks like this will impact readability.

My original design is to have an internal enum that contains the engine experimental features, and the description of each feature goes in a resource file called `ExperimentalFeatureDescription.resx`. When initializing the engine, it constructs a `ReadOnlyCollection<ExperimentalFeature>` out of the enum -- the name of an enum member is the experimental feature's name; read the description from the embedded resource. So when referencing an engine experimental feature in the engine code, you have to use the enum. It would serve the readability very well and also reduce errors like spelling a feature name incorrectly when directly using a string.

But I got feedback that (1) separating the description to a resource file is heavy given experimental features might not live long and we don't do resource string localization anyways; (2) having name and description together in the code is more readable. So I gave up the original design and turn to having an exp-feature author to manually add an instance for the read-only collection when adding a new exp-feature.

Now I again tend to have the enum, or a set of `const` string variables representing each exp-feature, so that we don't directly use literal strings when referencing a exp-feature in the engine.